File: packages/go/crudian/sql.go
package crudian

import (
	"fmt"
	"reflect"
)

type compiledWhere struct {
	SQL       string
	Args      []any
	NextIndex int // next 1-based placeholder index
}

func compileWhere(d Dialect, node WhereNode, startIndex int) (compiledWhere, error) {
	if startIndex < 1 {
		startIndex = 1
	}
	if node == nil {
		return compiledWhere{NextIndex: startIndex}, nil
	}
	switch n := node.(type) {
	case GroupNode:
		if len(n.Children) == 0 {
			return compiledWhere{NextIndex: startIndex}, nil
		}
		parts := make([]string, 0, len(n.Children))
		args := make([]any, 0)
		idx := startIndex
		for _, child := range n.Children {
			c, err := compileWhere(d, child, idx)
			if err != nil {
				return compiledWhere{}, err
			}
			if c.SQL == "" {
				continue
			}
			parts = append(parts, "("+c.SQL+")")
			args = append(args, c.Args...)
			idx = c.NextIndex
		}
		if len(parts) == 0 {
			return compiledWhere{NextIndex: startIndex}, nil
		}
		if len(parts) == 1 {
			sql := parts[0]
			return compiledWhere{SQL: sql[1 : len(sql)-1], Args: args, NextIndex: idx}, nil
		}
		join := " AND "
		if n.Type == "or" {
			join = " OR "
		}
		out := parts[0]
		for i := 1; i < len(parts); i++ {
			out += join + parts[i]
		}
		return compiledWhere{SQL: out, Args: args, NextIndex: idx}, nil
	case CondNode:
		col, err := AssertString(n.Column, "column")
		if err != nil {
			return compiledWhere{}, err
		}
		q := d.QuoteIdent(col)
		idx := startIndex
		switch n.Op {
		case OpEq:
			return compiledWhere{SQL: q + " = " + d.Placeholder(idx), Args: []any{n.Value}, NextIndex: idx + 1}, nil
		case OpNe:
			return compiledWhere{SQL: q + " <> " + d.Placeholder(idx), Args: []any{n.Value}, NextIndex: idx + 1}, nil
		case OpLt:
			return compiledWhere{SQL: q + " < " + d.Placeholder(idx), Args: []any{n.Value}, NextIndex: idx + 1}, nil
		case OpGt:
			return compiledWhere{SQL: q + " > " + d.Placeholder(idx), Args: []any{n.Value}, NextIndex: idx + 1}, nil
		case OpLte:
			return compiledWhere{SQL: q + " <= " + d.Placeholder(idx), Args: []any{n.Value}, NextIndex: idx + 1}, nil
		case OpGte:
			return compiledWhere{SQL: q + " >= " + d.Placeholder(idx), Args: []any{n.Value}, NextIndex: idx + 1}, nil
		case OpLike:
			return compiledWhere{SQL: q + " LIKE " + d.Placeholder(idx), Args: []any{n.Value}, NextIndex: idx + 1}, nil
		case OpIsNull:
			return compiledWhere{SQL: q + " IS NULL", Args: nil, NextIndex: idx}, nil
		case OpIsNotNull:
			return compiledWhere{SQL: q + " IS NOT NULL", Args: nil, NextIndex: idx}, nil
		case OpIn:
			vals, err := asSlice(n.Value)
			if err != nil {
				return compiledWhere{}, err
			}
			if len(vals) == 0 {
				return compiledWhere{}, NewError("in requires a non-empty array")
			}
			ph := make([]string, len(vals))
			for i := range vals {
				ph[i] = d.Placeholder(idx)
				idx++
			}
			return compiledWhere{
				SQL:       fmt.Sprintf("%s IN (%s)", q, joinComma(ph)),
				Args:      vals,
				NextIndex: idx,
			}, nil
		default:
			return compiledWhere{}, NewError("unknown op: " + string(n.Op))
		}
	default:
		return compiledWhere{}, NewError("invalid where node")
	}
}

func resolveWhere(w *WhereBuilder) WhereNode {
	if w == nil {
		return nil
	}
	return w.ToNode()
}

func asSlice(v any) ([]any, error) {
	if v == nil {
		return nil, NewError("in value must be an array")
	}
	rv := reflect.ValueOf(v)
	if rv.Kind() != reflect.Slice && rv.Kind() != reflect.Array {
		return nil, NewError("in value must be an array")
	}
	out := make([]any, rv.Len())
	for i := 0; i < rv.Len(); i++ {
		out[i] = rv.Index(i).Interface()
	}
	return out, nil
}

func joinComma(parts []string) string {
	if len(parts) == 0 {
		return ""
	}
	out := parts[0]
	for i := 1; i < len(parts); i++ {
		out += ", " + parts[i]
	}
	return out
}

func selectColumns(d Dialect, columns []string) string {
	if len(columns) == 0 {
		return "*"
	}
	parts := make([]string, len(columns))
	for i, c := range columns {
		parts[i] = d.QuoteIdent(c)
	}
	return joinComma(parts)
}

File: packages/js/src/sqlite/sql.ts
import {
  CrudianError,
  isWhereBuilder,
  type WhereInput,
  type WhereNode,
} from "../index.js"
import type { Dialect } from "../dialect/types.js"
import { sqliteDialect } from "../dialect/sqlite.js"

/** @deprecated Prefer dialect.quoteIdent; kept for local helper call sites. */
export function quoteIdent(name: string): string {
  return sqliteDialect.quoteIdent(name)
}

export function resolveWhere(input: WhereInput | undefined): WhereNode | undefined {
  if (input === undefined) return undefined
  if (isWhereBuilder(input)) return input.toNode()
  return input
}

export type CompiledWhere = {
  sql: string
  args: unknown[]
  /** Next 1-based placeholder index after this fragment. */
  nextIndex: number
}

/**
 * Compile a where AST using dialect placeholders.
 * @param startIndex 1-based index for the first placeholder in this fragment.
 */
export function compileWhere(
  dialect: Dialect,
  node: WhereNode | undefined,
  startIndex = 1,
): CompiledWhere {
  if (!node) return { sql: "", args: [], nextIndex: startIndex }

  if (node.type === "and" || node.type === "or") {
    if (node.children.length === 0) return { sql: "", args: [], nextIndex: startIndex }
    const parts: string[] = []
    const args: unknown[] = []
    let idx = startIndex
    for (const child of node.children) {
      const compiled = compileWhere(dialect, child, idx)
      if (!compiled.sql) continue
      parts.push(`(${compiled.sql})`)
      args.push(...compiled.args)
      idx = compiled.nextIndex
    }
    if (parts.length === 0) return { sql: "", args: [], nextIndex: startIndex }
    if (parts.length === 1) {
      return { sql: parts[0]!.slice(1, -1), args, nextIndex: idx }
    }
    const joiner = node.type === "and" ? " AND " : " OR "
    return { sql: parts.join(joiner), args, nextIndex: idx }
  }

  if (node.type !== "cond") {
    throw new CrudianError("invalid where node")
  }

  const col = dialect.quoteIdent(node.column)
  let idx = startIndex
  switch (node.op) {
    case "eq":
      return {
        sql: `${col} = ${dialect.placeholder(idx)}`,
        args: [node.value],
        nextIndex: idx + 1,
      }
    case "ne":
      return {
        sql: `${col} <> ${dialect.placeholder(idx)}`,
        args: [node.value],
        nextIndex: idx + 1,
      }
    case "lt":
      return {
        sql: `${col} < ${dialect.placeholder(idx)}`,
        args: [node.value],
        nextIndex: idx + 1,
      }
    case "gt":
      return {
        sql: `${col} > ${dialect.placeholder(idx)}`,
        args: [node.value],
        nextIndex: idx + 1,
      }
    case "lte":
      return {
        sql: `${col} <= ${dialect.placeholder(idx)}`,
        args: [node.value],
        nextIndex: idx + 1,
      }
    case "gte":
      return {
        sql: `${col} >= ${dialect.placeholder(idx)}`,
        args: [node.value],
        nextIndex: idx + 1,
      }
    case "like":
      return {
        sql: `${col} LIKE ${dialect.placeholder(idx)}`,
        args: [node.value],
        nextIndex: idx + 1,
      }
    case "isNull":
      return { sql: `${col} IS NULL`, args: [], nextIndex: idx }
    case "isNotNull":
      return { sql: `${col} IS NOT NULL`, args: [], nextIndex: idx }
    case "in": {
      const values = node.value
      if (!Array.isArray(values)) {
        throw new CrudianError("in value must be an array")
      }
      if (values.length === 0) {
        throw new CrudianError("in requires a non-empty array")
      }
      const placeholders = values.map(() => dialect.placeholder(idx++)).join(", ")
      return { sql: `${col} IN (${placeholders})`, args: values, nextIndex: idx }
    }
    default:
      throw new CrudianError(`unknown op: ${(node as { op: string }).op}`)
  }
}
